src/storage/sql_table.cpp

```diff
@@ -12,9 +11,226 @@ namespace terrier::storage {
+                             ->LogRecord::GetUnderlyingRecordBodyAs<RedoRecord>(),
+                 "This RedoRecord is not the most recent entry in the txn's RedoBuffer. Was StageWrite called "
+                 "immediately before?");
+  // TODO(Schema-Change): need to take care of tuple migration if the update touches the new columns added
+  //  The migration should be a delete (MVCC style) in old datatable followed by an insert in new datatable.
+
+  // get the version of current tuple slot
+  auto curr_tuple = redo->GetTupleSlot();
+  auto returned_slot = curr_tuple;
+
+  const auto tuple_version = curr_tuple.GetBlock()->data_table_->layout_version_;
+
+  TERRIER_ASSERT(tuple_version <= layout_version,
+                 "The iterator should not go to data tables with more recent version than the current transaction.");
+
+  bool result;
+  if (tuple_version == layout_version) {
+    result = tables_.at(layout_version).data_table_->Update(txn, curr_tuple, *(redo->Delta()));
+  } else {
+    // tuple in an older version, check if all modified columns are in the datatable version where the tuple is in
+
+    col_id_t ori_header[redo->Delta()->NumColumns()];
+    auto desired_v = tables_.at(layout_version);
+    auto tuple_v = tables_.at(tuple_version);
+    auto missing = AlignHeaderToVersion(redo->Delta(), tuple_v, desired_v, &ori_header[0]);
+
+    if (!missing) {
+      result = tuple_v.data_table_->Update(txn, redo->GetTupleSlot(), *(redo->Delta()));
+      std::memcpy(redo->Delta()->ColumnIds(), ori_header, sizeof(col_id_t) * redo->Delta()->NumColumns());
+    } else {
+      // touching columns that are in the desired schema, but not the actual schema
+      // do an delete followed by an insert
+
+      // get projected row from redo (This projection is deterministic for identical set of columns)
+      std::vector<col_id_t> col_ids;
+      for (const auto &it : desired_v.column_id_to_oid_map_) col_ids.emplace_back(it.first);
+      auto initializer = ProjectedRowInitializer::Create(desired_v.layout_, col_ids);
+      auto *const buffer = common::AllocationUtil::AllocateAligned(initializer.ProjectedRowSize());
+      auto *pr = initializer.InitializeRow(buffer);
+
+      // fill in values to the projection
+      result = Select(txn, curr_tuple, pr, layout_version);
+
+      if (result) {
+        // delete it from old datatable
+        result = tuple_v.data_table_->Delete(txn, curr_tuple);
+        if (result) {
+          // insert it to new datatable
+
+          // apply the change
+          StorageUtil::ApplyDelta(desired_v.layout_, *(redo->Delta()), pr);
+          const auto slot UNUSED_ATTRIBUTE = desired_v.data_table_->Insert(txn, *pr);
+        }
+      }
+      delete[] buffer;
+    }
+  }
+  if (!result) {
+    // For MVCC correctness, this txn must now abort for the GC to clean up the version chain in the DataTable
+    // correctly.
+    txn->SetMustAbort();
+  }
+  return std::make_pair(result, returned_slot);
```

Comment:
I believe that make_pair has very bad performance. Try using a struct instead.

Reply:
@kjobkjobanputra citation? Not that I don't believe you but I would interested to see what makes you think that.